In [1]:
import pandas as pd
import numpy as np

In [2]:
source_dir = '../data/undesired_outcomes_pred/'
dest_dir = source_dir
extension = '.csv'

In [3]:
elog = pd.read_csv(source_dir.__add__('elog.csv'))

In [4]:
pd.options.display.max_columns = 100

In [5]:
elog.sample(7)

Case ID                                     Activity  \
1267553  fc30182676dbb5f3          Control summary-Main-finish editing   
20253    e2df1e93cc8933f9                   Parcel document-Main-check   
57785    15034ad1b6fa4abd            Entitlement application-Main-save   
1506971  815e342d64754bcc    Payment application-Application-calculate   
1020528  066c7544a8d25996    Payment application-Application-calculate   
1969744  598ecac8b09c64c8  Geo parcel document-Reported-finish editing   
16167    42766e818831a58d                    Control summary-Main-save   

                 Resource       Complete Timestamp  (case) amount_applied0  \
1267553  Parcel automaton  2016-07-12 17:32:01.000                 1331.70   
20253              556da9  2015-08-03 16:39:31.000                24147.44   
57785              00037f  2016-10-27 16:40:46.349                62184.06   
1506971            727350  2016-10-13 16:30:24.313                 1458.98   
1020528            727350  2016-12-09 13:38:05.332                 1331.70   
1969744            727350  2017-12-03 05:35:02.224                25104.63   
16167              7d03a0  2015-11-04 10:03:35.000                 3734.20   

         (case) amount_applied1  (case) amount_applied2  \
1267553                     NaN                     NaN   
20253                       NaN                     NaN   
57785                  62295.36                     NaN   
1506971                     NaN                     NaN   
1020528                     NaN                     NaN   
1969744                     NaN                     NaN   
16167                       NaN                     NaN   

         (case) amount_applied3  (case) applicant  (case) area  \
1267553                     NaN  bef44be655372283       3.8414   
20253                       NaN  46e8fbfa218b90e8      78.1008   
57785                       NaN  c92481a07a616854     204.2089   
1506971                     NaN  ed71e131e3d49dda       4.1938   
1020528                     NaN  269b3818300e3a85       3.8414   
1969744                     NaN  ad64b8cce8765c8f      84.3800   
16167                       NaN  2ae7401d508af252      11.2921   

         (case) basic payment  (case) cross_compliance (case) department  \
1267553                  True                      0.0                e7   
20253                    True                      0.0                e7   
57785                    True                      3.0                6b   
1506971                  True                      0.0                4e   
1020528                  True                      0.0                4e   
1969744                  True                      0.0                e7   
16167                    True                      0.0                4e   

         (case) greening  (case) number_parcels  (case) payment_actual0  \
1267553             True                      3                 1331.70   
20253               True                     42                24062.87   
57785               True                     57                58519.18   
1506971             True                      2                 1444.84   
1020528             True                      9                 1331.70   
1969744             True                     24                25097.09   
16167               True                      9                 3733.57   

         (case) payment_actual1  (case) payment_actual2  \
1267553                     NaN                     NaN   
20253                       NaN                     NaN   
57785                    991.11                     NaN   
1506971                     NaN                     NaN   
1020528                     NaN                     NaN   
1969744                     NaN                     NaN   
16167                       NaN                     NaN   

         (case) payment_actual3  (case) penalty_ABP  (case) penalty_AGP  \
1267553                     NaN         

In [6]:
elog.loc[~pd.isnull(elog['(case) penalty_amount2'])]

Case ID                                           Activity  \
8383     b0d199e3fd21ffb4        Payment application-Application-mail income   
8384     b0d199e3fd21ffb4            Entitlement application-Main-mail valid   
8385     b0d199e3fd21ffb4         Payment application-Application-mail valid   
8386     b0d199e3fd21ffb4            Entitlement application-Main-mail valid   
8387     b0d199e3fd21ffb4                    Control summary-Main-initialize   
8388     b0d199e3fd21ffb4                 Control summary-Main-begin editing   
8389     b0d199e3fd21ffb4                Control summary-Main-finish editing   
8390     b0d199e3fd21ffb4                    Parcel document-Main-initialize   
8391     b0d199e3fd21ffb4                 Parcel document-Main-begin editing   
8392     b0d199e3fd21ffb4               Parcel document-Main-insert document   
8393     b0d199e3fd21ffb4                          Parcel document-Main-save   
8394     b0d199e3fd21ffb4                Parcel document-Main-finish editing   
8395     b0d199e3fd21ffb4                Reference alignment-Main-initialize   
8396     b0d199e3fd21ffb4                 Reference alignment-Main-performed   
8397     b0d199e3fd21ffb4          Department control parcels-Main-performed   
8398     b0d199e3fd21ffb4      Department control parcels-Main-begin editing   
8399     b0d199e3fd21ffb4            Entitlement application-Main-initialize   
8400     b0d199e3fd21ffb4                      Reference alignment-Main-save   
8401     b0d199e3fd21ffb4                          Parcel document-Main-save   
8402     b0d199e3fd21ffb4                Parcel document-Main-finish editing   
8403     b0d199e3fd21ffb4         Entitlement application-Main-begin editing   
8404     b0d199e3fd21ffb4             Entitlement application-Main-calculate   
8405     b0d199e3fd21ffb4      Department control parcels-Main-begin editing   
8406     b0d199e3fd21ffb4                      Reference alignment-Main-save   
8407     b0d199e3fd21ffb4      Department control parcels-Main-begin editing   
8408     b0d199e3fd21ffb4         Payment application-Application-initialize   
8409     b0d199e3fd21ffb4      Payment application-Application-begin editing   
8410     b0d199e3fd21ffb4          Payment application-Application-calculate   
8411     b0d199e3fd21ffb4             Entitlement application-Main-calculate   
8412     b0d199e3fd21ffb4             Entitlement application-Main-calculate   
...                   ...                                                ...   
1641626  f24f9e1ac0d7701f               Payment application-Objection-refuse   
1641627  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641628  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641629  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641630  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641631  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641632  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641633  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641634  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641635  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641636  f24f9e1ac0d7701f      Payment application-Objection-revoke approval   
1641637  f24f9e1ac0d7701f  Payment application-Objection-begin admissibil...   
1641638  f24f9e1ac0d7701f  Payment application-Objection-check admissibility   
1641639  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641640  f24f9e1ac0d7701f               Payment application-Objection-refuse   
1641641  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641642  f24f9e1ac0d7701f      Payment application-Objection-insert document   
1641643  f24f9e1ac0d7701f                      Payment application-Main-save   
1641644  f24f9e1ac

In [7]:
elog.loc[~pd.isnull(elog['(case) payment_actual3'])].sample(3)

Case ID                                   Activity Resource  \
1495000  6dcac1562d4ca766                    Inspection-On-Site-save   7850e8   
1036688  266b1d94f695c521  Payment application-Application-calculate   727350   
230670   99786dc412ea305c    Entitlement application-Main-mail valid      NaN   

              Complete Timestamp  (case) amount_applied0  \
1495000  2016-11-30 14:22:35.923                34769.12   
1036688  2016-10-26 16:26:59.823                42249.17   
230670   2015-05-13 00:00:00.000               150960.97   

         (case) amount_applied1  (case) amount_applied2  \
1495000                34794.71                35254.05   
1036688                42166.72                42166.72   
230670                150704.82               150656.52   

         (case) amount_applied3  (case) applicant  (case) area  \
1495000                35253.72  a5d52351e3091a8d     107.9741   
1036688                42166.72  54792368267e874c     131.1465   
230670                150697.10  17fe3fa90c1ceff4     398.9333   

         (case) basic payment  (case) cross_compliance (case) department  \
1495000                  True                      0.0                d4   
1036688                  True                      0.0                e7   
230670                   True                      0.0                d4   

         (case) greening  (case) number_parcels  (case) payment_actual0  \
1495000             True                     59                29492.92   
1036688             True                     25                42266.33   
230670              True                     57               150589.69   

         (case) payment_actual1  (case) payment_actual2  \
1495000                 5273.58                 -376.22   
1036688                    0.00                -1448.64   
230670                  -120.51                 -150.83   

         (case) payment_actual3  (case) penalty_ABP  (case) penalty_AGP  \
1495000                  -378.3               False               False   
1036688                     0.0               False               False   
230670                      0.0               False               False   

         (case) penalty_AJLP  (case) penalty_AUVP  (case) penalty_AVBP  \
1495000                False                False                False   
1036688                False                False                False   
230670                 False                False                False   

         (case) penalty_AVGP  (case) penalty_AVJLP  (case) penalty_AVUVP  \
1495000                False                 False                 False   
1036688                False                 False                 False   
230670                 False                 False                 False   

         (case) penalty_B16  (case) penalty_B2  (case) penalty_B3  \
1495000               False              False              False   
1036688               False              False              False   
230670                False               True              False   

         (case) penalty_B4  (case) penalty_B5  (case) penalty_B5F  \
1495000              False               True               False   
1036688              False              False               False   
230670               False              False               False   

         (case) penalty_B6  (case) penalty_BGK  (case) penalty_BGKV  \
1495000              False               False                False   
1036688              False               False                False   
230670               False               False                False   

         (case) penalty_BGP  (case) penalty_C16  (case) penalty_C4  \
1495000               False               False              False   
1036688               False               False              False   
230670                False               False              False   

         (case) penalty_C9  (case) penalty_CC  (case) penalty_GP1  \
1495000        

### Case lifecycle

In [8]:
len(set(elog['Case ID']))

43809

So there are 43809 cases present in the event log. Let's see how many of these cases are actually useful to use for training/testing/cross-validation.

In [9]:
elog.loc[elog['Activity'] == 'Payment application-Application-decide'].shape

(65437, 89)

In [10]:
len(set(elog.loc[elog['Activity'] == 'Payment application-Application-decide']['Case ID']))

43793

We see that there are only 43809 - 43793 = 16 cases that do not have the the 'Payment application-Application-decide' activity in their lifecycle (as far as can be deduced from the event logs)

Let's drop these 16 cases since they are incomplete in the context of this business question and are not that significant when one considers the number of cases at hand.

In [11]:
to_be_dropped_cases = set(elog['Case ID']).difference(set(set(elog.loc[elog['Activity'] == 'Payment application-Application-decide']['Case ID'])))

In [12]:
elog = elog[~elog['Case ID'].isin(to_be_dropped_cases)]

In [13]:
len(set(elog['Case ID'])) # for confirmation

43793

But since we want to be able to detect the undesired outcomes before the 'Payment application-Application-decide' the choice is made to consider the event that proceeds this activity in almost all the cases: 'Payment application-Application-finish editing'.

In [14]:
len(set(elog.loc[elog['Activity'] == 'Payment application-Application-finish editing']['Case ID']))

43793

In [15]:
elog_uo = elog[elog['Activity'] == 'Payment application-Application-finish editing']

### Undesired outcome #1

In [21]:
set(elog['Activity'])

{'Control summary-Main-begin editing',
 'Control summary-Main-change department',
 'Control summary-Main-finish editing',
 'Control summary-Main-initialize',
 'Control summary-Main-insert document',
 'Control summary-Main-save',
 'Control summary-Main-take original document',
 'Department control parcels-Main-begin editing',
 'Department control parcels-Main-change department',
 'Department control parcels-Main-insert document',
 'Department control parcels-Main-performed',
 'Department control parcels-Main-remove document',
 'Department control parcels-Main-save',
 'Entitlement application-Change-begin editing',
 'Entitlement application-Change-calculate',
 'Entitlement application-Change-change department',
 'Entitlement application-Change-decide',
 'Entitlement application-Change-finish editing',
 'Entitlement application-Change-initialize',
 'Entitlement application-Change-insert document',
 'Entitlement application-Change-remove document',
 'Entitlement application-Change-revoke d

In [28]:
def is_undesired_outcome1(x):
    case_id = x['Case ID']
    case_subset = elog[elog['Case ID'] == case_id]
    case_events = set(case_subset['Activity'])
    if 'Payment application-Application-begin payment' in case_events: # there has been a begin payment activity
        if 'Payment application-Application-abort payment' not in case_events: # no abort payment activity in the data
            return True
        else: # check whether it was before the end of the year
            earliest_date = min(list(case_subset['timestamp_year']))
            date_abort_payment = list(case_subset[case_subset['Activity'] == 
                                             'Payment application-Application-abort payment']['timestamp_year'])[0]
            if date_abort_payment > earliest_date:
                return True
    return False

In [29]:
elog_uo['undesired_outcome1'] = elog_uo.apply(is_undesired_outcome1, axis=1)

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [34]:
print("shape ", elog_uo.shape[0])
print("true " , elog_uo[elog_uo['undesired_outcome1'] == True].shape[0])
print("false " , elog_uo[elog_uo['undesired_outcome1'] == False].shape[0])

shape  138335
true  58843
false  79492


### Undesired outcome #2

In [35]:
def is_undesired_outcome2(x):
    case_id = x['Case ID']
    case_subset = elog[elog['Case ID'] == case_id]
    case_subprocesses = set(case_subset['subprocess'])
    if 'Change' in case_subprocesses or 'Objection' in case_subprocesses:
        return True
    else:
        return False

In [36]:
elog_uo['undesired_outcome2'] = elog_uo.apply(is_undesired_outcome2, axis=1)

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [39]:
elog_uo['undesired_outcome'] = np.where(elog_uo['undesired_outcome1'] | elog_uo['undesired_outcome2'], True, False)

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [ ]:
print("shape ", elog_uo.shape[0])
print("true " , elog_uo[elog_uo['undesired_outcome1'] == True].shape[0])
print("false " , elog_uo[elog_uo['undesired_outcome1'] == False].shape[0])

In [40]:
elog_uo.to_csv(dest_dir.__add__('elog_uo').__add__(extension), index=False)